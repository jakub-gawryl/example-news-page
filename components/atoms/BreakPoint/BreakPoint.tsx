import usePageSize from '../../../hooks/usePageResize';

type Props = {
  /** Min width (in px) */
  min?: number;

  /** Max width (in px) */
  max?: number;

  /** CSS display type for visible element (e.g. 'block' or 'flex') */
  display?: string;

  /** React children to display when screen size is between min and max */
  children?: React.ReactElement | React.ReactElement[];
};

const BreakPoint: React.FC<Props> = ({
  min = 0,
  max = 0,
  display='block',
  children
}) => {
  const {width} = usePageSize();
  const isMinVisible = width >= min;
  const isMaxVisible = max > 0 ? width <= max : true;
  const isVisible = isMinVisible && isMaxVisible;
  const cssDisplay = isVisible ? display : 'none';

  return (
    <div style={{display: cssDisplay}}>
      {children}
    </div>
  )
}

export default BreakPoint;
