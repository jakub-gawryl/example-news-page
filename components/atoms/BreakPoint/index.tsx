import BreakingPoint from './BreakPoint';

export default BreakingPoint;