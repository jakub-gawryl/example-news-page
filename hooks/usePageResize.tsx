import * as React from 'react';

const defaultSize = {
  width: 0,
  height: 0
};

const usePageResize = () => {
  const [size, setSize] = React.useState(defaultSize);

  React.useEffect(() => {
    const listenToResize = () => {
      const { width, height } = window.screen;
      setSize({
        width,
        height
      });
    };

    window.addEventListener('resize', listenToResize)
    listenToResize();
  }, [setSize]);

  return size;
}; 

export default usePageResize;