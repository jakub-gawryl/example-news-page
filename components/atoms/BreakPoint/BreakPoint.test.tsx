import BreakPoint from './index';
import { render } from '@testing-library/react';
import usePageResize from '../../../hooks/usePageResize';

jest.mock('../../../hooks/usePageResize', () => ({
  __esModule: true,
  default: jest.fn()
}));

const setMockPageWidth = (width: number) => (usePageResize as jest.Mock).mockReturnValue({width});

/*
  Scenarios for page width == 600
    - HIDE when min=601
    - SHOW when min=600
    - HIDE when max=599
    - SHOW when max=600
    - HIDE when min=200 & max=599
    - SHOW when min=600 & max=600 (testing also display='flex')
*/
describe('BreakPoint', () => {

  describe('for page width = 600', () => {
    beforeEach(() => setMockPageWidth(600));

    it('should SHOW component(s) with prop min=600', () => {
      const mockMin = 600;

      const {container} = render(
        <BreakPoint min={mockMin}>
          <p>Hello test!</p>
        </BreakPoint>
      );

      expect(container.querySelector('p')).toBeVisible();
    });

    it('should HIDE component(s) with prop min=601', () => {
      const mockMin = 601;

      const {container} = render(
        <BreakPoint min={mockMin}>
          <p>Hello test!</p>
        </BreakPoint>
      );

      expect(container.querySelector('p')).not.toBeVisible();
    });

    it('should SHOW component(s) with prop max=600', () => {
      const mockMax = 600;

      const {container} = render(
        <BreakPoint max={mockMax}>
          <p>Hello test!</p>
        </BreakPoint>
      );

      expect(container.querySelector('p')).toBeVisible();
    });

    it('should HIDE component(s) with prop max=599', () => {
      const mockMax = 599;

      const {container} = render(
        <BreakPoint max={mockMax}>
          <p>Hello test!</p>
        </BreakPoint>
      );

      expect(container.querySelector('p')).not.toBeVisible();
    });

    it('should HIDE component(s) with prop min=200 & max=599', () => {
      const mockMin = 200;
      const mockMax = 599;

      const {container} = render(
        <BreakPoint min={mockMin} max={mockMax}>
          <p>Hello test!</p>
        </BreakPoint>
      );

      expect(container.querySelector('p')).not.toBeVisible();
    });

    it('should SHOW component(s) with prop min=600 & max=600 & display=flex', () => {
      const mockMin = 600;
      const mockMax = 600;

      const {container} = render(
        <BreakPoint min={mockMin} max={mockMax} display='flex'>
          <p>Hello test!</p>
        </BreakPoint>
      );

      expect(container).toMatchSnapshot();
      expect(container.querySelector('p')).toBeVisible();
    });

  });
});