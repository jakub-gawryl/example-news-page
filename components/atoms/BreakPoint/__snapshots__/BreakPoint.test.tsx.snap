// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreakPoint for page width = 600 should SHOW component(s) with prop min=600 & max=600 & display=flex 1`] = `
<div>
  <div
    style="display: flex;"
  >
    <p>
      Hello test!
    </p>
  </div>
</div>
`;
